src/sinks/kafka/request_builder.rs

```diff
@@ -0,0 +1,120 @@
+use crate::config::log_schema;
+use crate::event::{Event, Finalizable, Value};
+use crate::internal_events::KafkaHeaderExtractionFailed;
+use crate::sinks::kafka::service::{KafkaRequest, KafkaRequestMetadata};
+use crate::sinks::util::RequestBuilder;
+use crate::template::{Template, TemplateRenderingError};
+use rdkafka::message::OwnedHeaders;
+
+pub struct KafkaRequestBuilder {
+    pub topic: Template,
+    pub key_field: Option<String>,
+    pub headers_field: Option<String>,
+}
+
+impl RequestBuilder<Event> for KafkaRequestBuilder {
+    type Metadata = KafkaRequestMetadata;
+    type Events = [Event; 1];
+    type Payload = Vec<u8>;
+    type Request = KafkaRequest;
+    type SplitError = TemplateRenderingError;
+
+    fn split_input(
+        &self,
+        mut event: Event,
+    ) -> Result<(Self::Metadata, Self::Events), Self::SplitError> {
+        let topic = self.topic.render_string(&event)?;
```

Comment:
Possibly neither here nor there for this PR but rendering something per-event always makes me wonder if we can cache the values in some way, or build a lookup structure. I have to imagine that for most deployments there's only going to be a handful of topics in play.

Reply:
Yea, that crossed my mind. I'll keep it in the back of my mind as I look at it. I don't think there's anything to do in this PR for it though.